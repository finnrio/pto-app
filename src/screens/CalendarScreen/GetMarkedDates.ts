function getDatesBetween(start: string, end: string) {
  const startDate = new Date(start);
  const endDate = new Date(end);
  const datesArray = [];

  startDate.setDate(startDate.getDate() + 1);

  while (startDate < endDate) {
    datesArray.push(startDate.toISOString().slice(0, 10));
    startDate.setDate(startDate.getDate() + 1);
  }
  return datesArray;
}

export default function GetMarkedDates(ptoEvents: any) {
  let markedDates: any = {};
  ptoEvents?.forEach((event: { start_date: string; end_date: string }) => {
    getDatesBetween(event.start_date, event.end_date).forEach((date) => {
      markedDates = {
        ...markedDates,
        [date.toString()]: { color: "#8ad4ff" },
      };
    });
    markedDates = {
      ...markedDates,
      [event.start_date]: { startingDay: true, color: "#8ad4ff" },
      [event.end_date]: { endingDay: true, color: "#8ad4ff" },
    };
  });
  return markedDates;
}
