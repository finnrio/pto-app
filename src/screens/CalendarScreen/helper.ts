function MergeEvents(currentEvents: any, newEvents: any) {
  const mergedEvents: any = {};

  Object.keys(currentEvents).forEach((date) => {
    mergedEvents[date] = currentEvents[date];
  });

  Object.keys(newEvents).forEach((date) => {
    if (mergedEvents[date]) {
      mergedEvents[date].dots.push(newEvents[date].dots[0]);
    } else {
      mergedEvents[date] = newEvents[date];
    }
  });

  return mergedEvents;
}

function getDatesBetween(start: string, end: string) {
  const startDate = new Date(start);
  const endDate = new Date(end);
  const datesArray = [];

  while (startDate <= endDate) {
    datesArray.push(startDate.toISOString().slice(0, 10));
    startDate.setDate(startDate.getDate() + 1);
  }
  return datesArray;
}

function GetDottedDatesFromEvents(events: any, key: any) {
  let markedDates: any = {};
  events?.forEach((event: { start_date: string; end_date: string }) => {
    getDatesBetween(event.start_date, event.end_date).forEach((date) => {
      markedDates = {
        ...markedDates,
        [date.toString()]: { dots: [key] },
      };
    });
  });
  return markedDates;
}

export { MergeEvents, GetDottedDatesFromEvents };
