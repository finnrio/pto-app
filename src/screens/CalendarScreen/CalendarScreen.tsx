import { Alert, View } from "react-native";
import React, { useState } from "react";
import { CalendarList } from "react-native-calendars";
import { MarkedDates } from "react-native-calendars/src/types";
import { useFocusEffect } from "@react-navigation/native";
import styles from "./styles";
import GetCurrentUserPTO from "../../firebase/firestore/GetCurrentUserPTO";
import GetMarkedDates from "./GetMarkedDates";
import GetCurrentUserData from "../../firebase/firestore/GetCurrentUserData";

export default function CalendarScreen({ navigation: { navigate } }: any) {
  const [startDate, setStartDate] = useState("");
  const [markedDates, setMarkedDates] = useState<MarkedDates>();
  const [managerId, setManagerId] = useState("");

  function getPTOEvents() {
    GetCurrentUserPTO().then((events) => {
      setMarkedDates(GetMarkedDates(events));
    });
  }

  function resetScreen() {
    setStartDate("");
    getPTOEvents();
    GetCurrentUserData().then((data) =>
      data.manager_id
        ? setManagerId(data.manager_id)
        : Alert.alert("Error", "Manager ID not found"),
    );
  }

  const startAlert = (date: string) =>
    Alert.alert(
      "PTO Request",
      `Create a PTO request starting on ${date}`,
      [
        {
          text: "Start Request",
          onPress: () => {
            setStartDate(date);
          },
          style: "default",
        },
        {
          text: "Cancel",
          style: "cancel",
        },
      ],
      { cancelable: true },
    );

  const endAlert = (date: string) =>
    Alert.alert(
      "PTO Request",
      `End PTO on ${date}`,
      [
        {
          text: "Complete Request",
          onPress: () => {
            navigate("PTO Request Form", {
              startDate,
              endDate: date,
              managerId,
            });
            resetScreen();
          },
          style: "default",
        },
        {
          text: "Cancel Request",
          onPress: () => {
            setStartDate("");
          },
          style: "destructive",
        },
        {
          text: "Cancel",
          style: "cancel",
        },
      ],
      { cancelable: true },
    );

  useFocusEffect(
    React.useCallback(() => {
      resetScreen();
    }, []),
  );

  return (
    <View style={styles.container}>
      <CalendarList
        onDayPress={(day) =>
          startDate ? endAlert(day.dateString) : startAlert(day.dateString)
        }
        markingType={"period"}
        markedDates={markedDates}
      />
    </View>
  );
}
