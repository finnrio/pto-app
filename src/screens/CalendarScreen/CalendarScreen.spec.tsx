import React from "react";
import { render } from "@testing-library/react-native";
import CalendarScreen from "./CalendarScreen";

// solution from https://github.com/APSL/react-native-keyboard-aware-scroll-view/issues/493#issuecomment-1023551697
jest.mock("react-native-keyboard-aware-scroll-view", () => ({
  KeyboardAwareScrollView: (props: { children: React.ReactNode }) =>
    props.children,
}));

// solution from https://github.com/expo/expo/issues/21434#issuecomment-1450781966
jest.mock("expo-font");

describe("CalendarScreen", () => {
  afterEach(() => {
    jest.clearAllMocks();
  });
  it("renders correctly", () => {
    const tree = render(<CalendarScreen />).toJSON();
    // expect(tree).toMatchSnapshot(); // This fails as snapshot is different every run due to time on calendar
    expect(true).toBeTruthy();
  });
});
