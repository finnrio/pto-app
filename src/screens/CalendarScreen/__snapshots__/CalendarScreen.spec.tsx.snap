// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CalendarScreen renders correctly 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flex": 1,
    }
  }
>
  <View
    style={
      {
        "flex": undefined,
      }
    }
  >
    <RCTScrollView
      data={
        [
          "2020-03-24T16:44:34.749Z",
          "2020-04-24T15:44:34.749Z",
          "2020-05-24T15:44:34.749Z",
          "2020-06-24T15:44:34.749Z",
          "2020-07-24T15:44:34.749Z",
          "2020-08-24T15:44:34.749Z",
          "2020-09-24T15:44:34.749Z",
          "2020-10-24T15:44:34.749Z",
          "2020-11-24T16:44:34.749Z",
          "2020-12-24T16:44:34.749Z",
          "2021-01-24T16:44:34.749Z",
          "2021-02-24T16:44:34.749Z",
          "2021-03-24T16:44:34.749Z",
          "2021-04-24T15:44:34.749Z",
          "2021-05-24T15:44:34.749Z",
          "2021-06-24T15:44:34.749Z",
          "2021-07-24T15:44:34.749Z",
          "2021-08-24T15:44:34.749Z",
          "2021-09-24T15:44:34.749Z",
          "2021-10-24T15:44:34.749Z",
          "2021-11-24T16:44:34.749Z",
          "2021-12-24T16:44:34.749Z",
          "2022-01-24T16:44:34.749Z",
          "2022-02-24T16:44:34.749Z",
          "2022-03-24T16:44:34.749Z",
          "2022-04-24T15:44:34.749Z",
          "2022-05-24T15:44:34.749Z",
          "2022-06-24T15:44:34.749Z",
          "2022-07-24T15:44:34.749Z",
          "2022-08-24T15:44:34.749Z",
          "2022-09-24T15:44:34.749Z",
          "2022-10-24T15:44:34.749Z",
          "2022-11-24T16:44:34.749Z",
          "2022-12-24T16:44:34.749Z",
          "2023-01-24T16:44:34.749Z",
          "2023-02-24T16:44:34.749Z",
          "2023-03-24T16:44:34.749Z",
          "2023-04-24T15:44:34.749Z",
          "2023-05-24T15:44:34.749Z",
          "2023-06-24T15:44:34.749Z",
          "2023-07-24T15:44:34.749Z",
          "2023-08-24T15:44:34.749Z",
          "2023-09-24T15:44:34.749Z",
          "2023-10-24T15:44:34.749Z",
          "2023-11-24T16:44:34.749Z",
          "2023-12-24T16:44:34.749Z",
          "2024-01-24T16:44:34.749Z",
          "2024-02-24T16:44:34.749Z",
          "2024-03-24T16:44:34.749Z",
          "2024-04-24T15:44:34.749Z",
          "2024-05-24T15:44:34.749Z",
          "2024-06-24T15:44:34.749Z",
          "2024-07-24T15:44:34.749Z",
          "2024-08-24T15:44:34.749Z",
          "2024-09-24T15:44:34.749Z",
          "2024-10-24T15:44:34.749Z",
          "2024-11-24T16:44:34.749Z",
          "2024-12-24T16:44:34.749Z",
          "2025-01-24T16:44:34.749Z",
          "2025-02-24T16:44:34.749Z",
          "2025-03-24T16:44:34.749Z",
          "2025-04-24T15:44:34.749Z",
          "2025-05-24T15:44:34.749Z",
          "2025-06-24T15:44:34.749Z",
          "2025-07-24T15:44:34.749Z",
          "2025-08-24T15:44:34.749Z",
          "2025-09-24T15:44:34.749Z",
          "2025-10-24T15:44:34.749Z",
          "2025-11-24T16:44:34.749Z",
          "2025-12-24T16:44:34.749Z",
          "2026-01-24T16:44:34.749Z",
          "2026-02-24T16:44:34.749Z",
          "2026-03-24T16:44:34.749Z",
          "2026-04-24T15:44:34.749Z",
          "2026-05-24T15:44:34.749Z",
          "2026-06-24T15:44:34.749Z",
          "2026-07-24T15:44:34.749Z",
          "2026-08-24T15:44:34.749Z",
          "2026-09-24T15:44:34.749Z",
          "2026-10-24T15:44:34.749Z",
          "2026-11-24T16:44:34.749Z",
          "2026-12-24T16:44:34.749Z",
          "2027-01-24T16:44:34.749Z",
          "2027-02-24T16:44:34.749Z",
          "2027-03-24T16:44:34.749Z",
          "2027-04-24T15:44:34.749Z",
          "2027-05-24T15:44:34.749Z",
          "2027-06-24T15:44:34.749Z",
          "2027-07-24T15:44:34.749Z",
          "2027-08-24T15:44:34.749Z",
          "2027-09-24T15:44:34.749Z",
          "2027-10-24T15:44:34.749Z",
          "2027-11-24T16:44:34.749Z",
          "2027-12-24T16:44:34.749Z",
          "2028-01-24T16:44:34.749Z",
          "2028-02-24T16:44:34.749Z",
          "2028-03-24T16:44:34.749Z",
          "2028-04-24T15:44:34.749Z",
          "2028-05-24T15:44:34.749Z",
          "2028-06-24T15:44:34.749Z",
          "2028-07-24T15:44:34.749Z",
        ]
      }
      getItem={[Function]}
      getItemCount={[Function]}
      getItemLayout={[Function]}
      horizontal={false}
      initialNumToRender={3}
      initialScrollIndex={50}
      keyExtractor={[Function]}
      nestedScrollEnabled={true}
      onContentSizeChange={[Function]}
      onLayout={[Function]}
      onMomentumScrollBegin={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEnabled={true}
      scrollEventThrottle={0.0001}
      scrollsToTop={false}
      showsHorizontalScrollIndicator={false}
      showsVerticalScrollIndicator={false}
      stickyHeaderIndices={[]}
      style={
        [
          {
            "backgroundColor": "white",
          },
          undefined,
        ]
      }
      testID="undefined.list"
      viewabilityConfigCallbackPairs={
        [
          {
            "onViewableItemsChanged": [Function],
            "viewabilityConfig": {
              "viewAreaCoveragePercentThreshold": 20,
            },
          },
        ]
      }
    >
      <View>
        <View
          style={
            {
              "height": 18000,
            }
          }
        />
        <View
          onFocusCapture={[Function]}
          style={null}
        >
          <Text
            style={
              [
                [
                  {
                    "minHeight": 360,
                    "width": 750,
                  },
                  {
                    "paddingLeft": 15,
                    "paddingRight": 15,
                  },
                  undefined,
                ],
                {
                  "color": "#2d4150",
                  "fontSize": 20,
                  "fontWeight": "200",
                },
              ]
            }
          >
            2024-05-24
          </Text>
        </View>
        <View
          onFocusCapture={[Function]}
          style={null}
        >
          <Text
            style={
              [
                [
                  {
                    "minHeight": 360,
                    "width": 750,
                  },
                  {
                    "paddingLeft": 15,
                    "paddingRight": 15,
                  },
                  undefined,
                ],
                {
                  "color": "#2d4150",
                  "fontSize": 20,
                  "fontWeight": "200",
                },
              ]
            }
          >
            2024-06-24
          </Text>
        </View>
        <View
          onFocusCapture={[Function]}
          style={null}
        >
          <Text
            style={
              [
                [
                  {
                    "minHeight": 360,
                    "width": 750,
                  },
                  {
                    "paddingLeft": 15,
                    "paddingRight": 15,
                  },
                  undefined,
                ],
                {
                  "color": "#2d4150",
                  "fontSize": 20,
                  "fontWeight": "200",
                },
              ]
            }
          >
            2024-07-24
          </Text>
        </View>
        <View
          style={
            {
              "height": 17280,
            }
          }
        />
      </View>
    </RCTScrollView>
  </View>
</View>
`;
